src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -104,3 +116,221 @@ def snapshot(self) -> Iterable[JobInfo]:
     def _try_fetch(self) -> Iterable[JobInfo]:
         for row in self._fetch(f"SELECT * FROM {self._schema}.{self._table}"):
             yield JobInfo(*row)
+
+
+def _dt_to_ms(dt: datetime):
+    epoch = datetime.utcfromtimestamp(0)
+    delta = dt - epoch
+    return int(delta.total_seconds() * 1000)
+
+
+def _generate_hourly_ranges(start_date: datetime, end_date: datetime) -> list[tuple[datetime, datetime]]:
+    # Ensure start_date is earlier than end_date
+    if start_date > end_date:
+        raise ValueError("Start date should be earlier than end date")
+    iteration_date = start_date
+    hourly_dates = []
+
+    while iteration_date <= end_date:
+        hourly_dates.append((iteration_date, iteration_date + timedelta(hours=1)))
+        iteration_date += timedelta(hours=1)
+
+    return hourly_dates
+
+
+def _get_job_cluster_from_task(task: RunTask) -> compute.ClusterSpec | None:
+    """
+    For submit runs, the following logic applies:
+    - if new_cluster is present, return it
+    - if existing_cluster_id is present, it means that the job was launched on all-purpose cluster
+    - since all-purpose clusters are not subject to the data_security_mode:None setting, we can skip them
+    - SubmitRuns do not support shared job clusters, so there is no need to check for them
+    """
+    if task.new_cluster is not None:
+        return task.new_cluster
+    return None
+
+
+@dataclass
+class SubmitRunInfo:
+    run_ids: str  # JSON-encoded list of run ids
+    hashed_id: str  # a pseudo id that combines all the hashable attributes of the run
+    failures: str = "[]"  # JSON-encoded list of failures
+
+
+_FS_LEVEL_KEY_REGEXP = [
+    "*.fs.s3a.*",
+    "*.fs.azure.*",
+    "*.fs.wasb.*",
+    "*.fs.abfs.*",
+    "*.fs.adl.*",
+]
+
+
+class SubmitRunsCrawler(CrawlerBase[SubmitRunInfo], JobsMixin, CheckClusterMixin):
+
+    def __init__(self, ws: WorkspaceClient, sbe: SqlBackend, schema):
```

Comment:
What is sbe?..

Reply:
What's the type of schema argument?

Reply:
no clue. We have **exactly the same code** in `JobsCrawler` and you didn't say a word (though you were reviewing), but here you're nitpicking. Any reason why?
```
    def __init__(self, ws: WorkspaceClient, sbe: SqlBackend, schema):
        super().__init__(sbe, "hive_metastore", schema, "jobs", JobInfo)
        self._ws = ws
```